### DISCUSSION TITLE
I cannot get bash completion for clang (installed through llvm) and pip3 (installed through python) on bash

### DISCUSSION THREAD

**Discussion by ghost** [Homebrew/discussions#1209] [Linux] [answered]
I cannot get bash completion for clang (installed through llvm) and pip3 (installed through python) on bash.
What do I do?

**Comment by gromgit** [comment]
Have you followed the instructions here? https://docs.brew.sh/Shell-Completion

**Comment by carlocab** [comment, marked as answer]
You need to configure your shell for these completions to work. You can add something like this to your `.{bash_}profile`:
```
eval $(pip3 completion --bash)
source $(brew --prefix llvm)/share/clang/bash-autocomplete.sh
```
Make sure `$HOMEBREW_PREFIX/bin` is in your `PATH` before these commands are run. (This is the default on Intel macOS.)

**Reply by ghost** [reply]
Yeah, that works.
And if I may ask, why aren't the completions for those programs automatically sourced by Homebrew like it does for formulae such as `the_silver_searcher`?
Also, is there some sort of list on programs whose completions I need to configure myself and how to do it?

**Reply by SMillerDev** [reply]
https://docs.brew.sh/Shell-Completion explains how to setup completion. If other formulae manage to do completion without this they might insert it if you run the program, or do some other magic.

**Reply by ghost** [reply]
I followed the instructions on that page a while back and It has been working perfectly. It's just that i've been stumbling upon programs whose completions don't work at all e.g. `fzf`

**Reply by carlocab** [reply]
Not all formulae install completions. There are many reasons why they don't:
1. Nobody has added them to the formula yet
2. The software just doesn't provide completions
3. The formula author intentionally left completions out of the formula (perhaps because they can be disruptive, so it's better for users to opt into them explicitly, for example)

**Reply by ghost** [reply]
I totally get you. I'm guessing this is the method that installs completions for a formula `bash_completion.install "<completions path>"`... if so, can I open a pull request for various formulae that haven't had the completions installed?

**Reply by SMillerDev** [reply]
Yes it is, and please do. Maybe you can also create the `zsh` and `fish` completion if available?

**Reply by carlocab** [reply]
Go ahead. However, it's probably not useful for keg-only formulae (e.g. `llvm`), and I believe `fzf` skips installing the completions on purpose. (The formula was contributed by the author of `fzf`.)

**Reply by ghost** [reply]
Cool
